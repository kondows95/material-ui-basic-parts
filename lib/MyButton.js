"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var core_1 = require("@material-ui/core");
var styles_1 = require("@material-ui/core/styles");
var styles_2 = require("./styles");
var useStyles = styles_1.makeStyles(function () { return ({
    progress: styles_2.getButtonLoadingStyle(),
}); });
var ignoreKeys = ['label', 'icon', 'loading'];
var MyButton = function (props) {
    var classes = useStyles();
    var loading = props.loading ? true : false;
    var copyProps = {};
    for (var _i = 0, _a = Object.keys(props); _i < _a.length; _i++) {
        var key = _a[_i];
        if (ignoreKeys.indexOf(key) === -1) {
            copyProps[key] = props[key];
        }
    }
    var progress = null;
    if (loading) {
        progress = react_1.default.createElement(core_1.CircularProgress, { size: 20, color: "primary", className: classes.progress });
    }
    var icon = null;
    if (props.icon) {
        icon = (react_1.default.createElement(core_1.Box, { display: "flex", alignItems: "center", justifyContent: "center", mr: 1 }, props.icon));
    }
    return (react_1.default.createElement(core_1.Button, __assign({}, copyProps, { disabled: loading, variant: "contained", fullWidth: true }),
        icon,
        props.label,
        progress));
};
exports.default = MyButton;
