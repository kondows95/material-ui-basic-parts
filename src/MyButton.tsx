import React from 'react';
import { Button, CircularProgress, Box } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { getButtonLoadingStyle } from './styles';

const useStyles = makeStyles(() => ({
    progress: getButtonLoadingStyle(),
}));

const ignoreKeys = ['label', 'icon', 'loading'];
type Props = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [key: string]: any;
    label: string;
    icon?: React.ReactElement;
    loading?: boolean;
};

type CopyProps = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [key: string]: any;
};

const MyButton: React.FC<Props> = (props) => {
    const classes = useStyles();

    const loading = props.loading ? true : false;
    const copyProps: CopyProps = {};
    for (const key of Object.keys(props)) {
        if (ignoreKeys.indexOf(key) === -1) {
            copyProps[key] = props[key];
        }
    }

    let progress = null;
    if (loading) {
        progress = <CircularProgress size={20} color="primary" className={classes.progress} />;
    }

    let icon = null;
    if (props.icon) {
        icon = <Box ml={1}>{props.icon}</Box>;
    }

    return (
        <Button {...copyProps} disabled={loading} variant="contained" fullWidth>
            {icon}
            {props.label}
            {progress}
        </Button>
    );
};
export default MyButton;
