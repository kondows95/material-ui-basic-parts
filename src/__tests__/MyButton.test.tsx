import React from 'react';
import TestLib from 'react-testing-library-wrapper';
import MyButton from '../lib/MyButton';
import { OpenInNew as OpenInNewIcon } from '@material-ui/icons';

let tLib: TestLib;

describe('Basic Usage', () => {
    beforeEach(() => {
        tLib = new TestLib(
            <MyButton label="MyLabel" icon={<OpenInNewIcon />} loading={true} className="MyClass" data-testid="myBtn" />
        );
    });

    it('You can add data-testid', () => {
        tLib.get('myBtn');
    });

    it('You can disable button and show CircularProgress', () => {
        const elem = tLib.get('myBtn') as HTMLButtonElement;
        expect(elem.disabled).toBeTruthy();
        expect(elem.innerHTML).toContain('CircularProgress');
    });

    it('You can set icon', () => {
        expect(tLib.get('myBtn').innerHTML).toContain('MuiSvgIcon');
    });

    it('You can add className', () => {
        expect(tLib.get('myBtn').getAttribute('class')).toContain('MyClass');
    });

    it('You can set button label', () => {
        expect(tLib.render.container.textContent).toBe('MyLabel');
    });

    it('Snapshot', () => {
        expect(tLib.render.asFragment()).toMatchSnapshot();
    });
});

describe('Expansion Usage', () => {
    beforeEach(() => {
        tLib = new TestLib(<MyButton label="MyLabel" data-testid="myBtn" />);
    });

    it('The Button is enable by default', () => {
        const elem = tLib.get('myBtn') as HTMLButtonElement;
        expect(elem.disabled).toBeFalsy();
        expect(elem.innerHTML).not.toContain('CircularProgress');
    });

    it('Icon is optional', () => {
        expect(tLib.get('myBtn').innerHTML).not.toContain('MuiSvgIcon');
    });
});
