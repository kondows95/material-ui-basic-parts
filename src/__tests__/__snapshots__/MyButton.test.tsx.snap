// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Usage Snapshot 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MyClass Mui-disabled MuiButton-fullWidth Mui-disabled"
    data-testid="myBtn"
    disabled=""
    tabindex="-1"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <div
        class="MuiBox-root MuiBox-root-315"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
          />
        </svg>
      </div>
      MyLabel
      <div
        class="MuiCircularProgress-root makeStyles-progress-281 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 20px; height: 20px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </span>
  </button>
</DocumentFragment>
`;
